### ISSUE TITLE
Future plans?

### ISSUE THREAD

**Issue by TFiFiE** [gcp/leela-zero#2157]
@gcp,

can you shed some light on any concrete plans for Leela Zero and whether we will finally see (intrusive) features that go beyond what DeepMind has done (if only in an experimental pipeline)? Right now, this project is in the exceptional position of being the single obvious choice for the general public to contribute computing resources to for the purpose of creating an as strong as possible go engine and it would be a real shame if the opportunity to try out things with such resources would be irreversibly lost if the training were to finally stall at 40x256 and competing projects that are more willing to experiment would start appearing and siphon off contributors.

**Comment by john45678** [comment]
LZ has been a great success in just over a year. I was wondering too if theres post 40b plans. (like trying to  produce stronger smaller nets or maybe a second run simultaneously to test ideas {like the LC0 guys are doing)).

**Comment by l1t1** [comment]
or study something inside the weights, such as the root of https://github.com/gcp/leela-zero/issues/2136

**Comment by Ishinoshita** [comment]
On what features go players may want to see, there was a quite detailed duscussion 6 month ago [here](https://github.com/gcp/leela-zero/issues/1758). I would personally like to see some work in the direction of handicap go/non-standard komi. In this respect, and as a preliminary, it might be useful to know what are plans for SAI. Could one day the community contribute to a 19x19 SAI, if need be ?

On methodology side, there are some fresh changes like recent -m 30, that might be worth to monitoring for a while to see the effect. And also doubts (at least questions) regarding proper gating and progress measurement (match games diversity issue). Such questions may need answers both before concluding current run or starting a new one. IMHO, current LZ has still a long way to go, although it is frustrating that it becomes harder an harder to gauge its real progress.

**Comment by Glrr** [comment]
> there are some fresh changes like recent -m 30
>
 Oh, I didn't notice that. It could explain why the selfplay games seems longer now.

**Comment by Vandertic** [comment]
> In this respect, and as a preliminary, it might be useful to know what are plans for SAI. Could one day the community contribute to a 19x19 SAI, if need be ?

Thanks @Ishinoshita! SAI will scale up to 19x19 as soon as we are ready, and then yeah, we will ask computing power to anyone willing to help.

With _ready_ I mean (a subset of) the following achievements:

- A clear understanding of the scaling-up policy to be used: we are wondering (and testing) if it is better to scale up from smaller to larger nets, as LZ project has done, or from lower to higher visits, while keeping a fixed net structure (say 20x256). This is because we fear that in the former case bad habits of the policy due to poor capacity of smaller nets may keep haunting larger nets.
- A robust way to evaluate the project progression, to see if everything is working and improvements are still coming. Gating is something but it is in no way robust. This is needed for example to decide when a scale-up is due.
- A way to deal with wrong/malicious data from clients.
- Some further tool to analyse train and match data to get useful insight.
- Update our code to LZ latest releases. (We forked a lot of time ago.)

We will ask computing power when we will be confident we won't waste it.

**Comment by Friday9i** [comment]
Humm, @wonderingabout, what to say?
- 50% gating: you strongly critize my "fancy" suggestion but you finish by saying you would pick either 50% or 55% gating. Sorry but aren't you needlessly critical?
- You also critize the ideas to generate more diversity in the opening of games as "needlessly intrusive": one of these idea has been implemented by Deepmind for AlphaZero (select moves within 1% winrate from the best move). Are you concluding that Deepmind experiments are "needless"?

Regarding simplicity, you are right, it's important. And that's good news: the 2 ideas just above (that you critized) are ... quite simple to implement : -)

Anyway, those are just ideas, it's up to @Vandertic to chose the ones he will implement: best wishes, waiting for news from your experiments!

**Comment by wonderingabout** [comment]
@Friday9i 

thanks for your kind reply
i wanted to point how overall i didnt see a clear direction and objectives in what you wanted to try : it felt like a mix of "let's try this and that and that oh and also this and that maybe, oh and maybe that too", mixed with a feel of wanting to rush things
correct me if i'm wrong, but it felt like this was more driven by worry (including worry that improvements are not fast or diversity is too small, etc..) than scientific method (where we fix one or 2 simple settings and stick with them good or bad until we can draw conclusions)
going with all these fancy experiments all at once, even though its interesting, wont make us able to conclude anything from it, it will just be a mix of many good and bad things
what deepmind did is good, but we dont have to treat it like the bible, more like take inspiration from it and do you own thing

then it's as you said, its up to SAI team to decide what they want to do, we are only sharing our views here
again, maybe we just have different approaches (straight versus spread experiments) so a different of opinion cant be helped, no bad intention with it

i'll update when i get results on my experiments, but its a long run thing :+1:

**Comment by Friday9i** [comment]
you're welcome @wonderingabout 
Just a few clarifications: 
- Regarding the "absence of clear direction": I must admit it was just a list of ideas and, as a list, it had indeed no single & clear direction, that was not the main point  ; -) 
- Regarding scientific method: I share your concern and am very interested to learn more about AI and to get more experiment results (so no difference between us on that point)! But IMHO, I get the feeling we are not progressing as fast as we could/should, which should encourage us to test new things. And unfortunately, for more than 4 months, we didn't test anything new (excepted the very recent and probably quite minor -m 30). Not sure it's the best way to understand new things ; -(
- Regarding my interest for the "fastest way to improve": I very much like this project, very interesting both theoretically (better understanding of AI) and practically (LZ is a wonderfulgo tool), but yes I admit I'd like it to improve faster (while still giving us many opportunities to better understand AI). Why? Because I invested in it (both time and significant money: a GTX1080 and months of electricity bill) and also because I think if LZ is getting crushed by other programs (PhoenixGo, MiniGo, etc...), participation will decrease and the project would die a slow death, which would really be a pity. It's not a race for sure, but to keep passionnate helpers on board, it's important to keep advancing at a good pace 

JMHO

**Comment by wonderingabout** [comment]
@Friday9i 

thanks for your introspection

i think this learning path on AI is similar to go : it's always about give and take

if you want to find fastest speed to improve, you are likely to lose on quality of the AI or other interesting areas
of course, no AI is yet perfect, so it is indeed possible to improve both speed and quality, but i think from an experimental point of view it is better to try one thing at a time, and if i am given the choice, i will always favour quality over speed
why ? because speed doesnt teach us anything, it is always possible to be faster with more hardware (add TPU or more GPU), or better software (add CUDA, cuDNN, tensorRT, etc.. instead of opencl), and other various ways
but the only way to know better about quality of the AI and AI settings is to try, try, try again, and fail, fail, fail again
i think leela zero is a quite rooted project now, no matter what we want to test here, it will be heavily determined by the already existing past, making it almost unchanged no matter what are the new experiments tried
as for quality, we learned a lot from leela zero, and i hope that it will improve too, however for human players it doesnt matter anymore because it is already superhuman, however for bot vs bot games leela zero has an average and plain middle game, had many blindspots, and its style is nowhere as thrilling to watch as phoenixgo or elf for example

remember, this is give and take : speed, or quality, not both (or a mix if you're undecided :+1: )

**Comment by Splee99** [comment]
One thing we couldn't ignore is that ELF is generally speaking still stronger in time parity. I think their weight structure with 224 inputs is interesting, because 15x15=225. 15 lines can be placed in the middle of 19 lines of the go board. LZ has 256 = 16x16 input. May be we can try 17x17 =289 input weight in the hope that we can deal with the symmetrical issue better.

**Comment by Ishinoshita** [comment]
@Splee99 Not sure I got the numerical argument there. ELF, like AZ (or LZ) , has **17** (or 18) input planes of 19x19 binary values, not '224 inputs' (or 256 for LZ). 

In ELF, 224 is **number of kernels** in each of the CNN layers, from the first one that processes the 17 input planes into 224 19x19 feature maps, to intermediate ones that process the 224 19x19 feature maps from previous layer into 224 19x19 new feature maps to feed the next layer. Each of the 224 kernels is a 3x3x224 tensor.

**Comment by Splee99** [comment]
@Ishinoshita OK, thanks for the explanation.

**Comment by gcp** [comment]
@TFiFiE 

I've been wanting to post something about this for a while, so I guess I sort-of have to now :-)

I'm very unlikely to make invasive changes to the current setup. This isn't because I don't believe trying to move ahead is not good, it's just that I know at this point I'm not going to manage to do that (and keep the current thing running in a good enough manner through any transition period!). This is both due to time restrictions, both total and continuous blocks (neither of which are getting better for me...), and motivational ones. I stated long ago in one of the issues that I expected 128x6 to be the last network and this to not be more than a proof of concept that the Alpha Zero method could be replicated in a distributed setting. With 256x40 networks that are >100 Elo ahead of the second ELF, we've gone, way, way, waaay beyond the original goals and the project has been much more successful than anticipated, thanks to many high quality contributions. But at this point, personally, I don't find trying to squeeze a few more (hundred?) Elo all that interesting, and I'm for example more concerned about the UI and GTP protocol situation. There's a fairly constant stream of people asking whether the "old" Leela will ever get updates, because Zero just doesn't work for them at this point, and my interest in algorithmic improvements is also exotic enough that it doesn't match up with what most of you are discussing.

So, practically, **I'll keep the current 256x40 running as is**, which probably has a (few?) month(s) to go with a last learning rate drop or playout increase, at least, but for the next leap ahead *someone else* (person or team) will have to step up and do the necessary work:

- Set a bunch of design decisions on what to address (strength? handicap?) and argue them well enough to convince people (or win by default 😀). For LZ, endless discussions could be cut short by referring to the AZ papers and copping out that way, but we sometimes still had very long arguments. This is very draining mentally and I'll be glad to let someone else call the shots and deal with that.

- Set up a server and training setup. For the record, zero.sjeng.org runs on a VPS that costs 5 EUR/month, and a basic workstation with a one time cost ~750 EUR GPU. For me, the main cost has always been the electricity bill, actually, because that's comparatively expensive here.

- Make the changes to the software to support what you want to test, or convince someone to do it for you. Better match setup? SAI style komi support? New network format? Improved training data format? You name it....

In terms of organization, it can be discussed here (though I find github issues kinda suck), on /r/cbaduk, on Discord (obviously not my choice, but it's not up to me!) or one could set up a new organization and transfer or fork the current repos into it. I want this project to continue to be a success as much as any of you, but realistically that won't be possible if I have to lead the "second generation". So someone else, person or team, has to step up and take up the lead.

**Comment by Hersmunch** [comment]
@gcp Fair call. Thanks for everything you've done!

p.s. "..and the lacking of continuous blocks needed.." taking you completely out of context :p https://arxiv.org/abs/1806.07366

**Comment by wonderingabout** [comment]
yes thats really reasonable

any project needs a director to guide people and have some authority, if no one can step up the project will eventually collapse, asking ourselves the right questions now is better than when its too late)

**Comment by TFiFiE** [comment]
@gcp, thanks for the reply.

Whatever transition will take place, I think it's very important that it should be as seamless as possible for the resource contributors in order to maintain as much of the ongoing inertia. My ideal vision would be that, by default, they remain part of a single pool that can be partitioned in any way by the project maintainers for the benefit of any or more branches (e.g. one main and one experimental one), but that, if they are explicit about it, individual contributors can always choose to restrict the branches their computing time can go to.

**Comment by tapsika** [comment]
> So, practically, I'll keep the current 256x40 running as is, which probably has a (few?) month(s) to go with a last learning rate drop or playout increase, at least

Just a minor note here. 15b nets trained on latest games are significantly stronger than what they seemed to peak at. This is because the 40b transition meant stronger nets, stronger games at a few times speed penalty. But there is another easy way to get significantly (even arbitrarily) stronger games at speed penalty...

**Comment by gcp** [comment]
Yes, that's more or less how it works with Leela Chess now, and that's good. We can just redirect zero.sjeng.org when the time comes, or a new client release could use another server address. There's been enough refactoring on AutoGTP by e.g. @Hersmunch that it's quite flexible at this point.

The worst outcome would be several Leela Zero branches with disparate features that can't easily be merged back. I'm going to reorganize a bit and put a (small) contributing & merging doc in the tree, and add a few more people as admin. So far I've been the only one to merge and push things, but that clearly doesn't scale and even less so in the future.

**Comment by gcp** [comment]
I set up https://github.com/leela-zero

Question now: do we FORK this repository into that org, or do we MOVE it?

I think a key difference is likely that moving would transfer all issues and releases?

**Comment by wonderingabout** [comment]
we do same as this emote

**Comment by wonderingabout** [comment]
P.S : no actually i have no idea what we should do and why lol

**Comment by alreadydone** [comment]
Keeping issues/releases would be nice. Is there any advantage of not moving? I guess you can first move and then fork to gcp/leela-zero to keep old links valid ...

**Comment by Hersmunch** [comment]
https://help.github.com/articles/transferring-a-repository/
Moving sounds mostly a good idea but here's the full list with some interesting bits highlighted by me.

> What's transferred with a repository?
> When you transfer a repository, **its issues, pull requests, wiki, stars, and watchers are also transferred. If the transferred repository contains webhooks, services, or deploy keys, they will remain associated after the transfer is complete. Git information about commits, including contributions, is preserved.** In addition:
> 
>- If the transferred repository is a fork, then it remains associated with the upstream repository.
>- If the transferred repository has any forks, then those **forks will remain associated with the repository after the transfer is complete.**
>-  If the transferred repository uses Git Large File Storage, all Git LFS objects are automatically moved. This transfer occurs in the background, so if you have a large number of Git LFS objects or if the Git LFS objects themselves are large, it may take some time for the transfer to occur. Before you transfer a repository that uses Git LFS, make sure the receiving account has enough data packs to store the Git LFS objects you'll be moving over. For more information on adding storage for personal accounts, see "Upgrading Git Large File Storage."
>-  When a repository is transferred between two user accounts, issue assignments are left intact. **When you transfer a repository from a user account to an organization, issues assigned to members in the organization remain intact, and all other issue assignees are cleared. Only owners in the organization are allowed to create new issue assignments.** When you transfer a repository from an organization to a user account, only issues assigned to the repository's owner are kept, and all other issue assignees are removed.
>-  If the transferred repository contains a GitHub Pages site, then links to the Git repository on the Web and through Git activity are redirected. However, we don't redirect GitHub Pages associated with the repository.
>- All links to the previous repository location are automatically redirected to the new location. When you use git clone, git fetch, or git push on a transferred repository, these commands will redirect to the new repository location or URL. However, to avoid confusion, we strongly recommend updating any existing local clones to point to the new repository URL.

**Comment by pangafu** [comment]
I'd like to see a stronger AI with like bellow:
1. The advance net : resnet is old, we could use resnext, senet for next.
2. The feather plan: I always believe the known algorithm in hard code is more efficient than deep learning.
3. Start RL from human go game, make a leela master to see the stength of ultima human.

**Comment by alreadydone** [comment]
"feather plan" means feature planes, I suppose ...

**Comment by pangafu** [comment]
@alreadydone yes~

**Comment by wonderingabout** [comment]
funny :laughing: (i am not making fun of you, it is just funny in a good way) :+1:

**Comment by anoek** [comment]
@gcp and all interested, speaking on behalf of online-go.com, we'd be happy to perpetually provide a sever and maintain / co-maintain it, as well as provide resources for training as well.

**Comment by gcp** [comment]
>I think we can make a interface that is easy to use even to people who are extremely unfamiliar with computer.

Sure, but that can happen entirely independently of Leela Zero work. About the only moment where it touches would be if GTP needs to be extended.

**Comment by gcp** [comment]
The repo was moved. Looks like github redirects nicely.

The most active contributors have admin access to the new org. I'll write up a small CONTRIBUTING document wrt. code style, dependencies, reviews, merging, etc.

**Comment by wonderingabout** [comment]
i guess i need to update the cloud tutorials with /leela-zero/leela-zero now

**Comment by gcp** [comment]
I wrote up contributor guidelines:
https://github.com/leela-zero/leela-zero/blob/56f639c62db78595bf61a91ff92e0bf0b931b0ef/CONTRIBUTING.md

**Comment by wonderingabout** [comment]
i really like this line

> If the answer to this question is no, then do not merge the code. Not merging a contribution (quickly) is annoying for the individual contributor. Merging a bad contribution is annoying for everyone who wants to contribute now and in the future.

**Comment by TFiFiE** [comment]
If he wants to be one, I heartily recommend @lightvector as one of the project admins.

One way in which I really would not like to see this project follow Leela Chess's footsteps, would be for important discussions and decisions to start taking place off-site, like on Discord. It would make it impossible for the casual observer to get a clear overview of what's going on. I challenge anyone to state in what way relaying the big stuff in the issues and pull requests here would be less than ideal. Even when real-time exchanges do produce fruitful stuff, it shouldn't be a problem then recapitulating them here.

**Comment by wonderingabout** [comment]
i just updated the cloud tutorials with the new repo owner

**Comment by gogre** [comment]
> @gcp I'm for example more concerned about the UI and GTP protocol situation

Quite rightly, for unless i am mistaken, leela-zero (whom i insist upon calling "Laizy" only because i like the name) is already strong enough to beat every human on the planet, so getting her to run even faster is only for the revhead birds who want to drag-race her against other gunslingers/streetracers - ie almost all the other contributors to this thread! :)

As to UI, whereas it would be nice for us non-geeks to have a free Christmas present we can open and start playing with immediately, that sort of thing is surely something that should **only** be undertaken by forkers and client bundlers such as Sabaki and Lizzie, because there is an analogy between Laizy and an OS - Laizy [OS] is the engine that does the work, and should be kept free from dependency upon of its client's UI [desktop, or whatever OS UIs are called these days].

Laizy is the dog, and eg Lizzie is its tail.  It's the tail's job to keep up with the dog, which can be made easier for tail developers if the dog speaks standardised GTP language in a standardised way, **_and that standard must always be backwards-compatible_**, unlike just about every flavour of Linux i've ever tried, which trip over their own feet everytime some geek decides to try to make it look more like Microsoft or whatever.

**Comment by alreadydone** [comment]
Once https://github.com/leela-zero/leela-zero/pull/1825 is merged, the training machine starts using the script in https://github.com/Hersmunch/leela-zero/pull/1, and leela-zero-server starts sending out the necessary GTP commands for komi and handicap https://github.com/leela-zero/leela-zero/pull/2096, we are ready to start komi+handicap training. Server can ask clients to do komi+handicap self-play with a small net, preferably a 15b which reacts reasonably to komi input; we can still use the games to train the main 40b net. Part of match games can be played with komi+handicap.
I think the moment Minigo 40b surpasses LZ 40b decisively will be a good occasion to make the switch (this may happen in the current run (v16) or the next run (v17 with SEnet) depending on when they decide to switch). We may or may not use the Minigo 40b nets to generate self-play games without handicap and with normal komi. Has anyone found a Minigo net that's good for dynamic komi?

**Comment by barrtgt** [comment]
Could you elaborate how this works? Is the black player playing weaker somehow? How do you have them play for the actual komi if you input a different komi?

**Comment by gcp** [comment]
>If he wants to be one, I heartily recommend @lightvector as one of the project admins.

Absolutely no disrespect to @lightvector but I don't see the point? He's never even contributed code here?

If you mean you'd like his input and opinions, then sure, go ahead and ask/have him tell. But you don't need to be admin of the github project to do that! That (mostly) means those people have write permissions in the github repository, and the people I made admin are those who contributed good quality code several times and have had review back-and-forths with me, so whom should have some idea what is OK to merge/pull and what needs more work. It's for the same reasons I wrote up #2186, in order to have some of this spelled out explicitly.

>One way in which I really would not like to see this project follow Leela Chess's footsteps, would be for 
>important discussions and decisions to start taking place off-site, like on Discord. It would make it 
>impossible for the casual observer to get a clear overview of what's going on

I am of similar point of view, and have avoided Discord because of the large distraction it potentially is. 

But if the People Who End Up Doing The Actual Work find it easier to collaborate that way, it's not up to me to say how they should organize themselves, IMHO.

**Comment by gcp** [comment]
It looks like moving did break Travis and Appveyor CI. Someone restored Travis yesterday, I tried to restore AppVeyor just now. Will have to check back if they're back to normal.

**Comment by gcp** [comment]
>It would make it impossible for the casual observer to get a clear overview of what's going on. I challenge anyone to state in what way relaying the big stuff in the issues and pull requests here would be less than ideal

It's probably better than Discord, but having these discussions in github issues is far from ideal either. You end up with a few of these mega-issues where all discussion happens and where it's pretty hard to find something back. They even default to folding past discussion at some point, so in-page search doesn't work very well either.

One area where github could use improvement, I think...

**Comment by l1t1** [comment]
>One area where github could use improvement, I think...
I saw some forum has a function of only display one authors post/reply. can github do it?

**Comment by gcp** [comment]
>Will have to check back if they're back to normal.

AppVeyor seems to be working again (had to update the webhook).

**Comment by roy7** [comment]
I turned on Travis for gtp2ogs this weekend and while in there saw it was off for the new leela-zero team, so I flipped that on too. ;)

**Comment by alreadydone** [comment]
Re: issue display and search improvements, I submitted a question through https://github.com/contact (linked to from https://github.community):
> Hi, do you have any plan to split issues into pages? That would help to save memory when expanding hidden comments in issues like https://github.com/gcp/leela-zero/issues/78 with 2000+ comments. Also, to find a comment in the middle, currently one has to click "Load more ..." many many times. It would be great if one can search within the issue and jump to the comment/the page in which the keyword is found. Thank you for your attention! 

and got the following reply:

> Re: Split issues into pages?
> Stacey Burns (GitHub Staff) <support@github.com> Fri, Jul 13, 2018, 5:45 AM
> Hi There,
> I'm not sure we have any plans to split Issue into pages but I'll pass the request onto our team!
> I can't promise if or when we'd add this but we'll make sure the request is in the right hands.
Cheers
Stacey

**Comment by alreadydone** [comment]
@barrtgt I am suggesting training with different komi values and different number of handicap stones, including with 7.5 komi and 0 handicap stones. The neural net so far only sees two komi values, namely 7.5 and -7.5. The hope is that when trained with games played under a wide variety of different komi values, the net would be able to generalize to all possible komi values. Training with handicap + appropriate komi provides such a wide variety of komi values.

I don't see why black plays (or will play) weaker. With appropriate komi to compensate for handicap stones, both sides must fight hard to win the game, and the winning probability should be close to 50%. You could start with white handicap stones and a large negative komi, but the neural net doesn't distinguish between black with a large positive komi and white with a large negative komi so you would just produce redundant training data.

There is indeed a risk of reducing play strength under 7.5 komi without handicap when the net no longer specializes on 7.5 komi. But if Minigo surpasses LZ, the play strength under 7.5 komi should no longer be our main pursuit, and we'd better enhance our play under different komi/handicap conditions. (Without training with different komi, they look unlikely to surpass us in handicap play.)

The way to input komi to the neural network is through the color planes (also known as stm(side-to-move) planes). It's already in the AlphaGo Zero paper:
> ... the colour feature C is necessary, because the komi is not observable.

This is tested to work well with some LZ 15b nets (see https://github.com/leela-zero/leela-zero/issues/1599) (that's why I suggest producing handicap games using them) but not for LZ 40b nets nor Minigo 20b/40b nets. So our 40b nets need further training to support different komi (which likely leads to better handicap play when combined with the dynamic komi approach).

**Comment by TFiFiE** [comment]
Perhaps the use of one or more [project boards](https://help.github.com/articles/about-project-boards/) would help keep things structured.

**Comment by barrtgt** [comment]
Thank you for explaining. I think this could make Leela much more useful for playing and analysis. How many komi values can our net support with this change?

**Comment by alreadydone** [comment]
I said
> The hope is that when trained with games played under a wide variety of different komi values, the net would be able to generalize to all possible komi values.

With 2 to 9 handicap stones, we should get a wide range of corresponding komi values for the net to interpolate between.

**Comment by TFiFiE** [comment]
Handicap games form just an arbitrary subset of those starting with a forced sequence of plays from one or both sides. Instead of passing, one might just as well force either or both to start out placing their first few stones in certain (suboptimal) patterns and (possibly) compensate that with appropriate komi.
I'd first start training with komi varying from 5.5 to 8.5 (as motivated in https://github.com/leela-zero/leela-zero/issues/1625#issuecomment-406738790) and see how well that gets generalized.

**Comment by Marcin1960** [comment]
I propose the following for the future.

I think that the best path would be to train smaller net. Not larger than ELF, but LITTLE larger than 15b , perhaps 17x192.

For several months most of resources went to the huge 40b net, in hope that the mere size will advance the project spectacularly. 

Clearly even 15b size did not exhaust its potential, a gradual increase to 17b or the like, will provide growth room, will be faster, more efficient and possibly exactly what should have been done before.

**Comment by herazul** [comment]
That won't help, we had new 15b because it was train on 40b self play games, which are on a way higher level than 15b self play games.
15b net won't improve from itself, the games it will play will be far inferior than 40b.

**Comment by Marcin1960** [comment]
@herazul: "That won't help, we had new 15b because it was train on 40b self play games"

There is no reason, not to train 17b (or something like that) on 40b self play games. Even more, that way the few months of huge net experimentation will not be wasted.

Otherwise, what do you propose? Keep running 40b in hope, something will change?

**Comment by herazul** [comment]
>There is no reason, not to train 17b (or something like that) on 40b self play games.

Yeah or a 20b, and it would be a pretty good net. But it would need the 40b net to keep training to have good 40b games to train on.

>Otherwise, what do you propose? Keep running 40b in hope, something will change?

That's the point : if you want good 15b or 20b net, you NEED the 40b to keep improving. It makes absolutely no sense to switch back to 15b or 20b in the official training process.

**Comment by Marcin1960** [comment]
@herazul: "There is no reason, not to train 17b (or something like that) on 40b self play games."
"Yeah or a 20b, and it would be a pretty good net. But it would need the 40b net to keep training to have good 40b games to train on."

Of course you need 40b to have 40b games. But perhaps you can have 17b or even 20b good games sooner and better.

Do you have any proof, besides a mere assumption, that nets smaller than 40b cannot be better than 40b?

My impression is that a huge jump from 20b to 40b was not a significant success.

**Comment by l1t1** [comment]
we can test best minigo 20x256 and 40x256 weights to judge which is stronger, both of them has trained for 20M games,(twice as lzs all self games , so lzs training is more efficient)
990 @ https://cloudygo.com/converted_model/v15-19x19/models/
930 @ https://cloudygo.com/converted_model/v16-19x19/models/

**Comment by herazul** [comment]
>>My impression is that a huge jump from 20b to 40b was not a significant success.

Well you have very high expectations then, because 40b nets are insanly strong. I don't really think you understand the achievement it is to have open source with community ressources nets that are that strong. The ONLY bot that may be stronger than LZ that we know of is fineart (and we aren't even sure because there hasn't been a tournament with lastest LZ nets.
And it's because the 40b nets got SO STRONG that 15b net could improve by training on 40b net self play games.

>Do you have any proof, besides a mere assumption, that nets smaller than 40b cannot be better than 40b?
I don't know for sure, the only thing i know is that by self-play, 40b net got way stronger than 15b or 20b, or even ELF, and is still improving.

**Comment by Marcin1960** [comment]
@herazul: "Well you have very high expectations then, because 40b nets are insanly strong."

So are smaller nets. Of course 40b is strongest now, because it was trained for months.

But does it mean that if, let 30b was given the same resources it could not be better?

My intuition (based on observation) is that the capacity of smaller nets between 15b and 40b was not exhausted, and that supposed extra capacity of 40b might be years from being utilized. 40b was chosen by an arbitrary doubling from 20b to 40b, based on mere expectation. The optimal size choice is in my opinion, somewhere between 15b and 40b, closer to the first than to the second.

**Comment by Friday9i** [comment]
@herazul "It makes absolutely no sense to switch back to 15b or 20b in the official training process": I totally agree that going back to 15b or 20b makes no sense, because the level of the games would be way lower and that would severly slow done progression... We have good @bjiyxo 15b nets thanks to the very deep games generated by the 40b nets!
But... there is a but!
40b selfplay games are much deeper than 15b games, but they are also much slower to generate, about 4x slower. And 15b nets trained with these games are about as strong as LZ203 on time parity. Hence, we could use theses 15b nets with ~4x more visits (ie ~6400 visits) for generating selfplay games:
- the level of these games should be comparable to the level of 40b games at 1600 visits
- the time to generate these games should be comparable

More generally, several ideas emerge from that:
- we could generate selfplay games with the recent 15b nets (from @bjiyxo and @bubblesld) at 6400 visits: theses games would be about as deep as 40b games (and would take about the same time to be generated), and they would add more diversity to the current selfplay games, which should help!
- we could also train 20b or 30b nets from recent selfplay games, and also use them to generate selfplay games, with even more diversity. That would be nice and fun to simultaneously generate selfplay games from 15b at 6400 visits, 25b at 3200 visits and 40b at 1600 visits, and to parallelize the training of these 15b, 25b and 40b nets ; -)! And when 15b nets become significantly less efficient (because they are too small and fundamentally limited), we should drop them and replace them by 60b or 80b nets for generating selfplay, while we could still try to train 15b nets and see if they improve a little bit!
- once progress slows down a lot with 40b at 1600 visits, we should go further the same way, either with 40b games and more visits (eg 3200 then 6400 visits) or larger nets (60b or 80b, or more filters?) at 1600 visits. From what we see, these very deep games (80b at 1600 or 40b at 3200 visits) should help push further the progression of the current 40b net once it stalls at 1600 visits.

**Comment by herazul** [comment]
> But does it mean that if, let 30b was given the same resources it could not be better?

Or maybe 36b would have been better. Or maybe 43b. Maybe even 154b would have work too. No one know, you have no data or reasoning behind your intuition, so instead of picking a random number of block that people liked because no reason, GCP picked 40b because : 
- it was the size used by AGZ, and it worked well for them.
- bjiyxo trained a 40b net at that time that was very good, so it showed even more that the size could work.


> smaller nets between 15b and 40b was not exhausted,

That doesn't really mean anything. We knew that 15b could have been stronger by training it on higher level of games (as that was proven true by training it on 40b self-play games, and ELF games before that). We don't know if it could have gotten stronger by it self.

@Friday9i i thought so too, but according to a discussion on minigo discord, andrew said that increasing the number of visit didn't increase the training by much if at all. So maybe the number of visit scaling is not very good, that would need to be tested before being implemented. What that maybe means is that a weaker elo a 4 times more visit might be worse to train on that a better and bigger net, even if they are the same elo at time parity. Hopefully minigo will test this more when they have a 40b net run without any problem soon, to see how it goes.

**Comment by Marcin1960** [comment]
@herazul: "But does it mean that if, let 30b was given the same resources it could not be better?
Or maybe 36b would have been better. Or maybe 43b. Maybe even 154b would have work too. "

No, 154b would NOT be better. It would so slow, that nobody would have a "good" computer to work on that.

Bigger is not always better. The reason why I propose 17b or something like that, is because it would be very FAST, most of people could participate in it, and since 15b did not exhaust its potential, certainly it must apply to 17b with even more!

**Comment by herazul** [comment]
@Marcin1960 
You need to understand something : it's not because 15b get better by training on 40b games that she could have gotten to that point by itself (and almost certainly sure that it couldn't have with our current training setup). The best way to have a strong 15b or 20b net is by training it on games generated by a stronger and bigger net.

Second thing is : it's not because you have a slow computer that a little net is better. Not at all. current 40b is perfectly fine on slow hardware.

**Comment by roy7** [comment]
For people who want smaller networks, maybe someone just needs to automate a training pipeline that will keep training networks of smaller sizes on the latest 40b games as they come out automatically. 

There's no reason we have to selfplay with 15b networks to keep improving the best 15b LZ based network available.

We aren't in the position of Minigo or LC0 to just start over from scratch every few weeks. Sure I am curious how we would do starting over with the modern LZ, but I think our best use of resources might be to keep pushing the 40b towards whatever strength cap it can attain and train smaller networks off those games.

**Comment by Marcin1960** [comment]
roy7: "There's no reason we have to selfplay with 15b networks"

Well, there is one good reason (I would rather increase the size a little to 17b):

to see if it develops another style, possibly opening new better ways of playing. Tying  smaller network to 40b games, prevents such success and discovery. 

Very important aspect - smaller net takes much less resources and experimenting with it will not devour much of resources and will be fast.

**Comment by barrtgt** [comment]
https://arxiv.org/pdf/1902.01917.pdf

**Comment by alreadydone** [comment]
@TFiFiE I agree that randomizing komi around the fair value is a good idea, since we see the nets don't interpolate very well between the trained komi 7.5 and -7.5. However, randomizing the opening moves might put too much burden on the net to learn many different openings. Handicap placements are some generally agreed upon starting positions, so they seem to be the most useful ones. But some servers/convention allows free handicap placement, and if we aim for that then randomizing the opening would be reasonable.

**Comment by nerai** [comment]
@Marcin1960 I think the misunderstanding is that what you really want are strong nets that work on small hardware, and not that we need to use these constraints all the time. I want to try and help resolve this.

I believe everyone here agrees that using 40b to train works *rather well*, and also has been shown by Deepmind to be viable. But not everyone is convinced that the same results could be achieved with smaller nets. You argue that they are faster (which is a fact), but there is evidence against the smaller nets being able to train beyond a certain limit by themselves. This is, after all, why there was a switch to 20b in the past; and it was even late - quite a few complained that we should have switched earlier! We also know Deepmind did switch to 40b when 20b stalled, and if I'm interpolating correctly it should have been not that much later than when LZ did. So there is a significant burden of proof to go back to smaller nets to train. I'm not saying it is impossible, but most people likely believe that it is at least less efficient than training 40b (including the speed penalty!).

Of course, we can always create new small nets from well trained new ones. It is nice that we learned that this works! So nobody prevents using future, better 40b nets to create better 15b nets, or 17b or 20b, for you to enjoy.

But regardless of that, a more fundamental question: Are we even sure that 40b is playing worse than 15b at low visits and time parity? As far as I'm aware, 40b might just be stronger even at something like 40b -v100 against 15b -v2000, which I believe is a scenario you are targeting. Have you got contrary results?

**Comment by alreadydone** [comment]
After moving, have to change `gcp` in url to `leela-zero` for the code to show:

With `leela-zero`:
https://github.com/leela-zero/leela-zero/blob/631b88fb73bbc9b45d9513af6ccaf8adfd4f8df9/src/Training.cpp#L241-L252

With `gcp`:
https://github.com/gcp/leela-zero/blob/631b88fb73bbc9b45d9513af6ccaf8adfd4f8df9/src/Training.cpp#L241-L252

**Comment by iopq** [comment]
>  As far as I'm aware, 40b might just be stronger even at something like 40b -v100 against 15b -v2000, which I believe is a scenario you are targeting. Have you got contrary results?

15b networks trained from recent data are competitive at time parity vs. 40b networks
